### PR TITLE
Add go to definition for required files

### PR DESCRIPTION
**Pull request by vinistock**
### Motivation

We offer autocompleting file paths for requires, so it'd be nice to allow users to also jump to the required file.

### Implementation

The implementation matches our approach for other listeners like hover. Currently, it's very focused only on allowing jumping to `require` and `require_relative`.

For `require`, we have to find the required file inside of the `$LOAD_PATH`. For `require_relative`, then we just need to turn the relative path into an absolute one to figure out the URI.

### Automated Tests

Added our standard structure for expectation testing.

### Manual Tests

1. Change the `definition_provider` to be `true` in `Executor#initialize_request`
2. Start the LSP
3. Go to `internal.rb`
4. Cmd Click to jump to required files

### CODE REVIEW

**Review by bitwise-aiden — APPROVED** [review]
Awesome feature!

For my own context, does the definition generation happen on hover, so if there is nothing to jump to, they won't have the right click option appear?

**Comment by andyw8** [comment]
Note: There is some kind of conflict happening with the NeoVim extension, which I showed to Vini. I'll try gather some more information, but it doesn't need to block this.

File: lib/ruby_lsp/requests/definition.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      private
+
+      sig { params(file: String).returns(T.nilable(String)) }
+      def find_file_in_load_path(file)
+        return unless file.include?("/")
+
+        # A hash of the format { "require/path/to/file.rb" => "/full/path/to/file.rb" }
+        available_paths = $LOAD_PATH.each_with_object({}) do |p, result|
+          Dir.glob("**/*.rb", base: p).each { |path| result[path] = "#{p}/#{path}" }
+        end
```
Comment (KaanOzkan) [inline lib/ruby_lsp/requests/definition.rb:89]: This is hash won't be small and it's created with every request. What do you think about testing it on core and seeing if there's a need to cache it? Things like the gem source key, value pairs could be easy wins for caching.

Not a blocker if this feature performs well on core.

File: lib/ruby_lsp/requests/definition.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    # The [definition
+    # request](https://microsoft.github.io/language-server-protocol/specification#textDocument_definition) jumps to the
+    # definition of the symbol under the cursor.
+    #
+    # Currently, only jumping to required files is supported.
+    #
+    # # Example
+    #
+    # ```ruby
+    # require "some_gem/file" # <- Jumping to the string definition will take you to the file
```
[the remark this answers is not in the document]
Comment (andyw8) [reply lib/ruby_lsp/requests/definition.rb:17]: That's somewhat specific to VS Code though, in vim/emacs I expect you could trigger it with a keybinding.

**Review by KaanOzkan — APPROVED** [review]

**Pull request merged**